### PR TITLE
Simplify case transaction index

### PR DESCRIPTION
**Pull request by emord**
`CaseTransaction` is our biggest table on ICDS. And this index is almost as big as the table (they're 48 GB and 46 GB respectively on commcarehq_p3). 

Following the original PR https://github.com/dimagi/commcare-hq/pull/11911 I think that this is only used in https://github.com/dimagi/commcare-hq/blob/4719481491718bf33c91acdc82ed5134cb62daa8/corehq/sql_accessors/sql_templates/case_has_transactions_since_sync.sql.

The benefit that we currently get from this is that that function is an index only scan. I'm curious if we think this is worth the cost to keep the index in memory. Because the synclog date and id are coming from the same synclog, the results rows are going to be roughly equivalent. The function uses EXIST and postgres might be smart enough to only pull a minimum of rows from disk before it returns with a True response.

Other possible improvements here:
- If `CaseTransaction` is expected to grow at a larger rate than other tables, we should consider partitioning it with architect(or pgpartman) or upgrading to pg10/11 and using native partitioning. The model is append only, so time would make sense to prioritize write, but it looks like all queries have case_id provided so case_id is probably a better choice.
- We throw away synclogs every 90 days. Should we also clean those up from the `CaseTransaction` table? Would the index and table be smaller if we updated those to NULL?
- synclog_id is [guaranteed to be a UUID ](https://github.com/dimagi/commcare-hq/blob/4719481491718bf33c91acdc82ed5134cb62daa8/corehq/ex-submodules/casexml/apps/phone/models.py#L413) since the sql migration, we should move the column from a char field to a UUID

### CODE REVIEW

**Comment by emord** [comment]
fwiw from postgres docs https://www.postgresql.org/docs/9.6/ddl-partitioning.html:

> The benefits will normally be worthwhile only when a table would otherwise be very large. The exact point at which a table will benefit from partitioning depends on the application, although a rule of thumb is that the size of the table should exceed the physical memory of the database server.

I don't know if "the size of the table" means just the table, or if it also includes the indexes, but our databases currently have 80 GB RAM.

Also wanted to explicitly note that the rollout of the index would need some more thought. Opening this PR mostly for discussion

**Comment by kaapstorm** [comment]
I don't have a lot of context on this, but this change seems like a good idea to me.

(Are you going to be creating a migration for this separately? Would you create a new index concurrently, and then drop the existing index?)

I support the idea of using the UUID data type for sync_log_id. [From the docs](https://www.postgresql.org/docs/9.1/datatype-uuid.html) it seems it could halve the size of the field, which should have a positive effect on the size of the index if you decide to keep sync_log_id in there.

It does seem right to set sync_log_id to NULL when its sync log is dropped. I'm not sure what effect that would have on table/index size, or performance considerations wrt the additional writes.

**Comment by snopoke** [comment]
Notes:
* the `case_has_transactions_since_sync` query is only used by `clean_owners` sync which ICDS does not use.
* agree that since we are using synclog date we likely don't need to have the synclog ID check as well
* agree that partitioning could be helpful - would need to audit the queries
* removing synclog after 90 days  - could be useful but not if we also remove it from the index
* I think the only usage we have is for that clean owners restore query. If we remove the need for it from there we could potentially drop the whole column

**Comment by emord** [comment]
Added two migrations for this index. Would need to be applied in separate release before deploy (which I can handle).

> the case_has_transactions_since_sync query is only used by clean_owners sync which ICDS does not use.
> I think the only usage we have is for that clean owners restore query. If we remove the need for it from there we could potentially drop the whole column

I found another use of this index is also used in a livequery restore, so I don't think that's possible

> removing synclog after 90 days - could be useful but not if we also remove it from the index

I think it would still be useful for reducing the overall table size though. Its hard to say on the overall impact, but if we end up doing the pg_repack work for online vacuuming, those null columns would repack more efficiently than they currently would

**Review by kaapstorm — APPROVED** [review]

File: corehq/form_processor/migrations/0093_remove_old_transaction_index.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.22 on 2019-07-20 19:16
+from __future__ import absolute_import, unicode_literals
+
+from django.db import migrations
+
+TABLE_NAME = 'form_processor_casetransaction'
+INDEX_NAME = 'form_processor_casetrans_case_id_server_date_sync_f7e3e655_idx'
+COLUMNS = ('case_id', 'server_date')
```
Comment (emord) [inline corehq/form_processor/migrations/0093_remove_old_transaction_index.py]:
```suggestion
COLUMNS = ('case_id', 'server_date', 'sync_log_id')
```

**Comment by emord** [comment]
I am planning to come back to this once we split our shards on ICDS

**Pull request closed without merging**